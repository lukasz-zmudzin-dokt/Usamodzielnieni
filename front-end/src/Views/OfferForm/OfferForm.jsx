import React, { useState, useContext, useEffect } from "react";
import { registerLocale } from "react-datepicker";
import { Form, Container, Card, Button, Row, Alert } from "react-bootstrap";
import { voivodeships } from "constants/voivodeships";
import FormGroup from "components/FormGroup";
import { sendData, getSelects } from "Views/OfferForm/functions/fetchData";
import { UserContext } from "context";
import polish from "date-fns/locale/pl";
import { useHistory } from "react-router-dom";

registerLocale("pl", polish);

const OfferForm = () => {
  const history = useHistory();
  const [validated, setValidated] = useState(false);
  const [fail, setFail] = useState(false);
  const [arrays, setArrays] = useState({});
  const [disabled, setDisabled] = useState(false);

  const [offer, setOffer] = useState({
    offer_name: "",
    company_name: "",
    company_address: "",
    voivodeship: voivodeships[0],
    description: "",
    expiration_date: "",
    category: "",
    type: "",
    pay_from: 12,
    pay_to: 12,
    pay_period: "",
    currency: "",
    pay_valid: true
  });

  const context = useContext(UserContext);

  useEffect(() => {
    setDisabled(true);
    const loadSelects = async (token) => {
      let res;
      try {
        res = await getSelects(token);
      } catch (e) {
        console.log(e);
        setFail(true);
        res = { categories: [], types: [] };
      }
      setArrays(res);
      setOffer({
        offer_name: "",
        company_name: context.data.company_name,
        company_address: context.data.company_address,
        voivodeship: voivodeships[0],
        description: "",
        expiration_date: "",
        category: res.categories[0],
        type: res.types[0],
      });
      setDisabled(false);
    };
    loadSelects(context.token);
  }, [context.data.company_address, context.data.company_name, context.token]);

  const submit = (event) => {
    const form = event.currentTarget;
    event.preventDefault();
    if (form.checkValidity() === false) {
      event.stopPropagation();
    } else {
      setDisabled(true);
      const year = expiration_date.getFullYear();
      const month =
        expiration_date.getMonth() + 1 < 10
          ? `0${expiration_date.getMonth() + 1}`
          : expiration_date.getMonth() + 1;
      const day =
        expiration_date.getDate() < 10
          ? `0${expiration_date.getDate()}`
          : expiration_date.getDate();
      const newDate = `${year}-${month}-${day}`;
      sendData({ ...offer, expiration_date: newDate }, context.token)
        .then(() => {
          history.push("/myOffers");
        })
        .catch(() => {
          setFail(true);
          setDisabled(false);
        });
    }
    setValidated(true);
  };

  // const checkPayRangeValidity = () => {
  //   if (pay_from > pay_to) {
  //     setOffer({
  //       pay_valid: false
  //     });
  //     console.log("ŹLE");
  //   } else {
  //     setOffer({
  //       pay_valid: true
  //     });
  //   }
  // };

  const {
    offer_name,
    company_address,
    company_name,
    description,
    expiration_date,
    voivodeship,
    category,
    type,
    pay_from,
    pay_to,
    currency,
    pay_period,
    pay_valid
  } = offer;

  return (
    <Container className="offerForm">
      <Card className="offerForm__card">
        <Card.Header as="h2" className="offerForm__header">
          Dodaj ofertę
        </Card.Header>
        <Card.Body>
          <Form
            onSubmit={submit}
            noValidate
            validated={validated}
            className="offerForm__form"
          >
            <div className="offerForm__wrapper">
              <FormGroup
                header="Nazwa stanowiska"
                setVal={(val) => setOffer({ ...offer, offer_name: val })}
                val={offer_name}
                incorrect="Podaj nazwę stanowiska"
                length={{ min: 1, max: 50 }}
                id="offer_name"
                required
              />
              <FormGroup
                header="Nazwa firmy"
                id="company_name"
                setVal={(val) => setOffer({ ...offer, company_name: val })}
                val={company_name}
                incorrect="Podaj nazwę firmy"
                length={{ min: 1, max: 70 }}
                required
                disabled
              />
              <FormGroup
                header="Adres firmy"
                id="company_address"
                setVal={(val) => setOffer({ ...offer, company_address: val })}
                val={company_address}
                incorrect="Podaj lokalizację"
                length={{ min: 1, max: 200 }}
                required
                disabled
              />
              <FormGroup
                header="Województwo"
                id="voivodeship"
                array={voivodeships}
                type="select"
                setVal={(val) => setOffer({ ...offer, voivodeship: val })}
                val={voivodeship}
                required
              />
              <FormGroup
                header="Wymiar pracy"
                id="type"
                setVal={(val) => setOffer({ ...offer, type: val })}
                val={type}
                type="select"
                array={arrays.types}
                required
                incorrect="Podaj wymiar pracy np. staż,praca"
              />
              <FormGroup
                header="Wynagrodzenie od:"
                id="pay_from"
                type="text"
                setVal={(val) => setOffer({ ...offer, pay_from: val })}
                val={pay_from}
                required
                // onChange={checkPayRangeValidity()}
                // setCustomValidity={`${pay_valid} ? 'Dobrze' : 'Źle'`}
              />
              <FormGroup
                header="Okres wypłaty wynagrodznia"
                id="pay_period"
                type="select"
                array={["za godzinę", "za dzień", "za miesiąc", "za rok", "jednokrotnie"]}
                setVal={(val) => setOffer({ ...offer, pay_period: val })}
                val={pay_period}
                required
              />
            </div>
            <div className="offerForm__wrapper">
              <FormGroup
                header="Opis stanowiska"
                id="description"
                type="textarea"
                setVal={(val) => setOffer({ ...offer, description: val })}
                val={description}
                incorrect="Podaj opis"
                length={{ min: 1, max: 1000 }}
                required
              />
              <FormGroup
                header="Branża"
                id="category"
                setVal={(val) => setOffer({ ...offer, category: val })}
                val={category}
                type="select"
                array={arrays.categories}
                required
                incorrect="Podaj branżę np. IT, marketing"
              />
              <FormGroup
                header="Ważne do:"
                id="expiration_date"
                type="date"
                setVal={(val) => setOffer({ ...offer, expiration_date: val })}
                val={expiration_date}
                required
              />
              <FormGroup
                header="Wynagrodzenie do:"
                id="pay_to"
                type="text"
                setVal={(val) => setOffer({ ...offer, pay_to: val })}
                val={pay_to}
                //onChange={checkPayRangeValidity()}
                required
              />
              <FormGroup
                header="Waluta wynagrodzenia:"
                id="currency"
                type="select"
                array={["PLN", "EUR", "USD"]}
                setVal={(val) => setOffer({ ...offer, currency: val })}
                val={currency}
                required
              />
            </div>
            {fail === true ? (
              <Row className="w-100 justify-content-center align-items-center m-0">
                <Alert variant="danger">
                  Coś poszło nie tak. Spróbuj ponownie póżniej.
                </Alert>
              </Row>
            ) : null}
            <Row className="w-100 justify-content-center align-items-center m-0">
              <Button
                variant="primary"
                type="submit"
                className=""
                disabled={disabled}
              >
                {disabled ? "Ładowanie..." : "Dodaj"}
              </Button>
            </Row>
          </Form>
        </Card.Body>
      </Card>
    </Container >
  );
};

export default OfferForm;
