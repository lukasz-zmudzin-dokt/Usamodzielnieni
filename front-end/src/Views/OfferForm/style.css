.offerForm {
  min-height: 100vh;
  padding-top: 100px;
}

.react-datepicker-wrapper {
  width: 100% !important;
}

@media (min-width: 768px) {
  .offerForm {
    display: flex;
    align-items: center;
    justify-content: center;
    padding-top: 0px;
  }
  .offerForm__card {
    margin-top: 50px;
    width: 100% !important;
  }
  .offerForm__form {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
  }
  .offerForm__wrapper {
    display: flex;
    flex-direction: column;
    width: 48%;
    justify-content: center;
  }
  .offerForm__textContainer {
    flex-grow: 1 !important;
    display: flex;
    flex-direction: column;
  }
  .offerForm__textarea {
    flex-grow: 1;
  }
}
