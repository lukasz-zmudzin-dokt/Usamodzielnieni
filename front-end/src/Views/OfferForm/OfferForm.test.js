import React from "react";
import {
  render,
  fireEvent,
  waitForElement,
  wait
} from "@testing-library/react";
import { Router } from "react-router-dom";
import { MemoryRouter } from "react-router-dom";
import OfferForm from "Views/OfferForm";
import { createMemoryHistory } from "history";

const renderWithRouter = (
  ui,
  {
    route = "/offerForm",
    history = createMemoryHistory({ initialEntries: [route] })
  } = {}
) => {
  return {
    ...render(<Router history={history}>{ui}</Router>),
    history
  };
};

describe("OfferForm", () => {
  let failFetch;
  beforeAll(() => {
    global.fetch = jest.fn().mockImplementation((input, init) => {
      return new Promise((resolve, reject) => {
        if (failFetch) {
          resolve({ status: 500 });
        }
        switch (init.method) {
          case "POST":
            resolve({ status: 200 });
            break;
          default:
            reject({});
            break;
        }
      });
    });
  });

  beforeEach(() => {
    failFetch = false;
    jest.clearAllMocks();
  });

  it("renders correctly", () => {
    const { container } = render(
      <MemoryRouter>
        <OfferForm />
      </MemoryRouter>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should clear state when offer is send", async () => {
    const { getByPlaceholderText, getByTestId, getByLabelText } = render(
      <MemoryRouter>
        <OfferForm />
      </MemoryRouter>
    );

    fireEvent.change(getByPlaceholderText("Nazwa stanowiska"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByPlaceholderText("Nazwa firmy"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByPlaceholderText("Adres firmy"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByLabelText("Województwo"), {
      target: { value: "lubelskie" }
    });
    fireEvent.change(getByLabelText("Opis stanowiska"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByLabelText("Ważne do:"), {
      target: {
        value: new Date()
      }
    });

    fireEvent.click(getByTestId("submitBtn"));

    await waitForElement(() => getByLabelText("Ważne do:"));

    expect(getByPlaceholderText("Nazwa stanowiska").value).toBe("");
    expect(getByPlaceholderText("Nazwa firmy").value).toBe("");
    expect(getByPlaceholderText("Adres firmy").value).toBe("");
    expect(getByLabelText("Województwo").value).toBe("dolnośląskie");
    expect(getByLabelText("Opis stanowiska").value).toBe("");
    expect(getByLabelText("Ważne do:").value).toBe("");
  });
  it("should not clear state when api return failure", async () => {
    failFetch = true;

    const { getByPlaceholderText, getByTestId, getByLabelText } = render(
      <MemoryRouter>
        <OfferForm />
      </MemoryRouter>
    );

    fireEvent.change(getByPlaceholderText("Nazwa stanowiska"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByPlaceholderText("Nazwa firmy"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByPlaceholderText("Adres firmy"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByLabelText("Województwo"), {
      target: { value: "lubelskie" }
    });
    fireEvent.change(getByLabelText("Opis stanowiska"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByLabelText("Ważne do:"), {
      target: {
        value: new Date()
      }
    });

    fireEvent.click(getByTestId("submitBtn"));

    await waitForElement(() => getByTestId("fail"));

    expect(getByTestId("fail")).toBeInTheDocument();
  });

  it("should not use fetch when form isn't validated", async () => {
    const { getByPlaceholderText, getByTestId, getByLabelText } = render(
      <MemoryRouter>
        <OfferForm />
      </MemoryRouter>
    );

    fireEvent.change(getByPlaceholderText("Nazwa stanowiska"), {
      target: { value: "" }
    });
    fireEvent.change(getByPlaceholderText("Nazwa firmy"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByPlaceholderText("Adres firmy"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByLabelText("Województwo"), {
      target: { value: "lubelskie" }
    });
    fireEvent.change(getByLabelText("Opis stanowiska"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByLabelText("Ważne do:"), {
      target: {
        value: ""
      }
    });

    fireEvent.click(getByTestId("submitBtn"));

    expect(fetch).toHaveBeenCalledTimes(0);
  });

  it("should redirect when offer is send", async () => {
    const { history, getByLabelText, getByTestId } = renderWithRouter(
      <OfferForm />
    );

    fireEvent.change(getByLabelText("Nazwa stanowiska"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByLabelText("Nazwa firmy"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByLabelText("Adres firmy"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByLabelText("Województwo"), {
      target: { value: "lubelskie" }
    });
    fireEvent.change(getByLabelText("Opis stanowiska"), {
      target: { value: "abcd" }
    });
    fireEvent.change(getByLabelText("Ważne do:"), {
      target: {
        value:
          "Wed Dec 04 2020 00:00:00 GMT+0100 (czas środkowoeuropejski standardowy)"
      }
    });

    fireEvent.click(getByTestId("submitBtn"));

    await wait(() => {
      expect(fetch).toHaveBeenCalled();
    });
    expect(history.location.pathname).toEqual("/myOffers");
  });
});
