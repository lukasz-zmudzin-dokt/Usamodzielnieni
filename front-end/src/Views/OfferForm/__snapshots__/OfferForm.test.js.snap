// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferForm renders correctly 1`] = `
<div
  class="offerForm container"
>
  <div
    class="offerForm__card card"
  >
    <h2
      class="offerForm__header card-header"
    >
      Dodaj ofertę
    </h2>
    <div
      class="card-body"
    >
      <form
        class="offerForm__form"
        novalidate=""
      >
        <div
          class="offerForm__wrapper"
        >
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="offer_name"
            >
              Nazwa stanowiska
            </label>
            <input
              class="form-control"
              data-testid="default"
              id="offer_name"
              maxlength="50"
              minlength="1"
              placeholder="Nazwa stanowiska"
              required=""
              type="text"
              value=""
            />
            <div
              class="invalid-feedback"
            >
              Podaj nazwę stanowiska
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="company_name"
            >
              Nazwa firmy
            </label>
            <input
              class="form-control"
              data-testid="default"
              disabled=""
              id="company_name"
              maxlength="70"
              minlength="1"
              placeholder="Nazwa firmy"
              required=""
              type="text"
              value="abc"
            />
            <div
              class="invalid-feedback"
            >
              Podaj nazwę firmy
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="company_address"
            >
              Adres firmy
            </label>
            <input
              class="form-control"
              data-testid="default"
              disabled=""
              id="company_address"
              maxlength="200"
              minlength="1"
              placeholder="Adres firmy"
              required=""
              type="text"
              value="ul. def 1, 00-000 abc"
            />
            <div
              class="invalid-feedback"
            >
              Podaj lokalizację
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="voivodeship"
            >
              Województwo
            </label>
            <select
              class="form-control"
              id="voivodeship"
              required=""
            >
              <option>
                dolnośląskie
              </option>
              <option>
                kujawsko-pomorskie
              </option>
              <option>
                lubelskie
              </option>
              <option>
                lubuskie
              </option>
              <option>
                łódzkie
              </option>
              <option>
                małopolskie
              </option>
              <option>
                mazowieckie
              </option>
              <option>
                opolskie
              </option>
              <option>
                podkarpackie
              </option>
              <option>
                podlaskie
              </option>
              <option>
                pomorskie
              </option>
              <option>
                śląskie
              </option>
              <option>
                świętokrzyskie
              </option>
              <option>
                warmińsko-mazurskie
              </option>
              <option>
                wielkopolskie
              </option>
              <option>
                zachodniopomorskie
              </option>
            </select>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="salary_min"
            >
              Wynagrodzenie od (zł / miesiąc)
            </label>
            <input
              class="form-control"
              data-testid="default"
              id="salary_min"
              max="-1"
              min="0"
              placeholder="Wynagrodzenie od (zł / miesiąc)"
              required=""
              type="number"
              value=""
            />
            <div
              class="invalid-feedback"
            >
              Pole musi być mniejsze od maksymalnej stawki
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="salary_max"
            >
              Wynagrodzenie do (zł / miesiąc)
            </label>
            <input
              class="form-control"
              data-testid="default"
              id="salary_max"
              max="999999"
              min="1"
              placeholder="Wynagrodzenie do (zł / miesiąc)"
              required=""
              type="number"
              value=""
            />
            <div
              class="invalid-feedback"
            >
              Pole musi być większe od minimalnej stawki
            </div>
          </div>
        </div>
        <div
          class="offerForm__wrapper"
        >
          <div
            class="textContainer form-group"
          >
            <label
              class="form-label"
              for="description"
            >
              Opis stanowiska
            </label>
            <textarea
              class="offerForm__textarea form-control"
              id="description"
              maxlength="1000"
              minlength="1"
              placeholder="Wpisz treść... "
              required=""
            />
            <div
              class="invalid-feedback"
            >
              Podaj opis
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="category"
            >
              Branża
            </label>
            <select
              class="form-control"
              id="category"
              required=""
            >
              <option>
                xd
              </option>
            </select>
            <div
              class="invalid-feedback"
            >
              Podaj branżę np. IT, marketing
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="type"
            >
              Wymiar pracy
            </label>
            <select
              class="form-control"
              id="type"
              required=""
            >
              <option>
                IT
              </option>
            </select>
          </div>
          <div
            class="form-group"
          >
            <label
              class="form-label"
              for="expiration_date"
            >
              Ważne do
            </label>
            <div
              class="align-items-center m-0 form-row"
            >
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container"
                >
                  <input
                    class="form-control"
                    id="expiration_date"
                    placeholder="Ważne do"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="w-100 justify-content-center align-items-center m-0 row"
        >
          <button
            class="btn btn-primary"
            data-testid="submitBtn"
            type="submit"
          >
            Dodaj
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
